integritee-cli: separate into library and binary

n-prat opened the pull request:
Hello,

Following up on https://github.com/integritee-network/worker/issues/1167 I have started a basic refactoring of the cli.
I needed to use `integritee-cli` from Android(and soon iOs) as a library so I cherry-picked and cleanup the required changes from our current branch and I am proposing this refactor.

NOTE: For this first draft, I aimed for the "minimal changeset" and not really for the "cleanest lib interface"... 
I am open for feedback!

Overview of the changes:
- most code from main.rs moved to lib.rs; with added structs `CliResultOk`, `CliError`, `CliResult`
- all `*Commands` return `CliResult`
- most of the logic changes are in `cli/src/trusted_operation.rs`
	- `perform_trusted_operation` now return new struct `TrustedOpResult`
	- I tried to correctly identify the previous "None means an error" vs "None means there was nothing to return"

TBD:
- do you want me to "go deeper" refactoring and expose a "proper lib"
  ie expose all `Subcommand` directly as `pub fn`[ideally keeping the `integritee-cli` as-is to not break anything]
- I could also in the PR ensure the cli works on Android; to do this two options: patch `sp-wasm-interface` and remove deps on `wasmi`+`wasmtime`; or patch `wasmtime` to allow compiling on Android+Emulator
  - For this option, not sure how to implement it; in my project I have a [dirty "patch"](https://github.com/Interstellar-Network/wallet-app/compare/master...cli-refactor-without-wasm) but ideally it should be done in this repo instead

clangenb commented:
Thanks for the PR, I am very sorry that I did not have the time yet to have a look at it. I hope to find time next week!

nathanprat commented:
No rush! This is not blocking for me.
Thanks for the update.

clangenb commented on cli/src/lib.rs:
```diff
@@ -0,0 +1,103 @@
+	command: Commands,
+}
+
+pub enum CliResultOk {
```
I somehow have issues with the name, but I can't come up with a better one. :)

clangenb reviewed (commented):
Thanks again for your contribution!

Sorry, for coming back so late to this! I must say that I very much like the approach and I think we don't need many further adjustments for this to merge it!

I just have minor remarks and a question. I would ask you to make this PR up-to-date, so we can run it in our CI. :)

clangenb commented:
Are you going to pick this up again? We are actually now having a business interest in this PR too, so if you don't work on this soonish, do you mind if someone from our team takes this task?

nathanprat commented on cli/src/lib.rs:
```diff
@@ -0,0 +1,103 @@
+	None,
+}
+
+#[derive(Debug, Snafu)]
```
Right. I used to default to `snafu` everywhere b/c `thiserror` did not support `no_std` but:
- I believe it can work now since `error_in_core`
- It is not really relevant for this repo/cli anyway

OverOrion replied:
Sorry for chiming in uninvited, but I think this might be useful here:
- There is an open PR in `thiserror` for making it `no_std` compatible: https://github.com/dtolnay/thiserror/pull/211
- In the meantime the PR author decided to submit `thiserror-core` which supports `no_std`: https://crates.io/crates/thiserror-core

clangenb replied:
As we are talking about the CLI anyhow, we don't care about no-std in this particular case :)

I don't know yet, how we want to handle the error in our no-std domain. I think, I will just wait and see towards which solution the ecosystem gravitates.

nathanprat commented:
I am definitely still interested in this feature. 
But I can't work on it right now. Next week is doable. 
Would that work for you? Else I don't mind if someone from your team wants to take over.

About the process: would you be ok with a rebase+force push to get up to date?
Or rather a merge?

nathanprat commented:
I have just squashed+rebased on master to clean up the dozen of commits I had on this branch.

Still some work to done about your comments (snafu to remove, cleanup, etc.) and I have probably left some `TODO`.
Also need to add some basic tests.

clangenb reviewed (changes requested):
Thanks for the update. I have approved your PR now as a test because we are trying to make our CI run for external contributors. We have recently migrated to self-hosted runners, and we don't know yet, how we can enable it to you.

OverOrion commented:
Hey @nathanprat,

Could you please add an empty commit to this PR? I believe it will help the CI get out of the _stuck_ state.

nathanprat commented:
Done.

Note that I had a `[ci skip]` in the previous commit so that could explain things... My bad.

n-prat commented:
Quick commits to fix the issues raised by CI :
- update in `cli/src/evm/commands/`; `cargo clippy --features evm -- -D warnings` now passes
- "taplo": `taplo fmt --check` now passes

clangenb commented on cli/tests/basic_tests.rs:
```diff
@@ -0,0 +1,26 @@
+
+	let commands = commands::match_command(&cli);
+	assert!(commands.is_ok());
+}
```
Thanks for adding these tests!

clangenb reviewed (commented):
Thanks for the fixes!

I think we now have a state that is ready to merge, or do you have any open points?

If not, I would like to ask you to merge master into your branch and we merge your PR afterwards.

n-prat commented:
I still need to handle what you said about indentation(I'm doing it now). While I'm at it, I will rebase to get it up to date.

The only open point I see is the `TODO` on  `CliResultOk::None`: I put it here as a placeholder/reminder that some functions are not properly exposed in the cli.
Ideally it should be removed and all the calls to eg `cmd.run(cli)` should directly return a `CliResult`.
But I am fine with merging it that way now, and down the line the needs arise for new variants, it will be a quick PR away.

n-prat commented:
- updated: `get_layer_two_evm_nonce` and `get_layer_two_nonce` with your remark about indentation
- updated: cli/tests/basic_tests.rs: tests were not passing
- rebased branch to get it up-to-date

Still open:
- cli/tests/basic_tests.rs is NOT called from CI; I don't really understand how you CI is structured so I am going some pointers if you want me to call these in CI
- cf "TODO on CliResultOk::None"

clangenb commented on cli/src/commands.rs:
```diff
@@ -43,12 +43,19 @@ pub enum Commands {
 
-pub fn match_command(cli: &Cli) {
+pub fn match_command(cli: &Cli) -> CliResult {
+	#[allow(non_snake_case, unused_variables)]
```
Why do we have this here?

n-prat replied:
No idea... Good catch!

clangenb commented:
Hi @n-prat,

Thanks for the update! I can say though that your tests are being executed. The snippet below is from the insanely long log of the `Build Worker and Cargo test` stage of the last [run](https://github.com/integritee-network/worker/actions/runs/4926741022/jobs/8806155813?pr=1183). However, I agree that we should simplify our CI to make this more obvious, which I think should be very much possible:

```
 #16 351.9 test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
#16 351.9 
#16 351.9      Running tests/basic_tests.rs (target/release/deps/basic_tests-e0bbeb8dbde7f650)
#16 351.9 
#16 351.9 running 2 tests
#16 351.9 test test_version ... ok
#16 351.9 test test_help ... ok
```

The question is now, these tests should have been called before too, which means that the CI might not catch when certain tests failed, see https://github.com/integritee-network/worker/issues/1299, but investigation should not be done in your PR.

That being said, I think I am fine with the return values for now, and I would favor a follow-up PR to improve the current code.

I just have the small question from the above comment, otherwise, I think we are ready to merge it!

n-prat commented:
I had forgotten to replace `Snafu` by `thiserror`.
Should be good now!

n-prat commented:
Same! Thanks for merging this!
Cheers
